pagefaultgames/pokerogue
# [Move] Implement Alluring Voice and Burning Jealousy

## What are the changes?
Implementing the special functionality of Alluring Voice and Burning Jealousy, which will either confuse or burn target Pokemon if they have had their stats boosted previously on the same turn.

## Why am I doing these changes?
Both moves are partially implemented and I think they're cool.

## What did change?
* Added `STATS_BOOSTED` tag. This is added to a Pokemon until the end of the current turn when they recieve a stat boost.
* Added `ConfuseIfBoostedAttr` class for Alluring Voice, which will confuse the target if their stats have been boosted this turn.
* Added `BurnIfBoostedAttr` class for Burning Jealousy, which will burn the target(s) if their stats have been boosted this turn.

I had previously attempted to implement stat boost tracking using a `PokemonTurnData` property, but found problems when it came to boosts that occured on summon on turn 1. The `resetTurnData()` function call within `TurnInitPhase` would reset the attribute before move selection. There are other ways this could be done, but I have found using a battler tag instead provides a cleaner solution. I'm not entirely sure if it fits with the intended use case of battler tags, though.

Another thing to note: Stat boost copying, such as by Ditto's transform or psych up, do not trigger the move effects (verified on showdown).

### Screenshots/Videos
Example of the move effects triggering on turn 1:

[2024-05-29 19-44-16.webm](https://github.com/pagefaultgames/pokerogue/assets/33560264/be8f076c-2f8d-41af-876a-b64b2db00b38)

[2024-05-29 19-45-01.webm](https://github.com/pagefaultgames/pokerogue/assets/33560264/07df7d27-6196-4ce1-be17-24b5f20afd5b)

Example of the move effects triggering after the target has boosted their stats:

[2024-05-29 19-51-21.webm](https://github.com/pagefaultgames/pokerogue/assets/33560264/7024fff4-6ece-4b37-871d-71c1f175f2ea)

Example of move effects not triggering, for good measure:

[2024-05-29 19-52-43.webm](https://github.com/pagefaultgames/pokerogue/assets/33560264/f81348e7-ec4a-409b-840d-5470d1a3f93b)


## How to test the changes?




I just made use of overrides, using abilities that boost stats on summon and stat boosting moves.
The effects trigger if the moves hit after stats have been boosted, so your starter will need to be slower than the opponent in most situations.
I skipped to the Eternatus battle to make sure it works against the boosts bosses recieve when one of their health bars is knocked out.

## Checklist
